File: lib/features/admin/services/group_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/group_model.dart';
import '../models/user_model.dart';

class GroupService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  CollectionReference get _groupsCollection => _firestore.collection('groups');
  CollectionReference get _usersCollection => _firestore.collection('users');

  // Create a new group
  Future<String> createGroup(String name, GroupLevel level, String? adminId) async {
    final docRef = await _groupsCollection.add({
      'name': name,
      'level': level.toString(),
      'adminId': adminId,
      'memberIds': [],
      'createdAt': FieldValue.serverTimestamp(),
    });
    return docRef.id;
  }

  // Stream all groups (for Main Admin)
  Stream<List<GroupModel>> getGroupsStream() {
    return _groupsCollection.snapshots().map((snapshot) {
      return snapshot.docs.map((doc) => GroupModel.fromFirestore(doc)).toList();
    });
  }

  // Stream groups for a specific admin (for Group Admin)
  Stream<List<GroupModel>> getGroupsByAdmin(String adminId) {
    return _groupsCollection
        .where('adminId', isEqualTo: adminId)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => GroupModel.fromFirestore(doc)).toList();
    });
  }

  // Add a member to a group
  Future<void> addMemberToGroup(String groupId, String userId) async {
    final batch = _firestore.batch();

    // 1. Update group members list
    batch.update(_groupsCollection.doc(groupId), {
      'memberIds': FieldValue.arrayUnion([userId]),
    });

    // 2. Update user's assigned group ID
    batch.update(_usersCollection.doc(userId), {
      'assignedGroupId': groupId,
    });

    await batch.commit();
  }

  // Remove a member from a group
  Future<void> removeMemberFromGroup(String groupId, String userId) async {
    final batch = _firestore.batch();

    // 1. Update group members list
    batch.update(_groupsCollection.doc(groupId), {
      'memberIds': FieldValue.arrayRemove([userId]),
    });

    // 2. Clear user's assigned group ID
    batch.update(_usersCollection.doc(userId), {
      'assignedGroupId': null,
    });

    await batch.commit();
  }

  // Get members of a specific group
  Stream<List<UserModel>> getGroupMembers(String groupId) {
    return _usersCollection
        .where('assignedGroupId', isEqualTo: groupId)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => UserModel.fromFirestore(doc)).toList();
    });
  }

  // Delete a group
  Future<void> deleteGroup(String groupId, List<String> memberIds) async {
    final batch = _firestore.batch();

    // 1. Reset assignedGroupId for all members
    for (var userId in memberIds) {
      batch.update(_usersCollection.doc(userId), {'assignedGroupId': null});
    }

    // 2. Delete the group document
    batch.delete(_groupsCollection.doc(groupId));

    await batch.commit();
  }
}

File: web-react/src/core/services/GroupService.js
import { db } from "../../lib/firebase";
import {
    collection,
    addDoc,
    updateDoc,
    deleteDoc,
    doc,
    query,
    where,
    getDocs,
    onSnapshot,
    serverTimestamp,
    arrayUnion,
    arrayRemove,
    or
} from "firebase/firestore";

export const GroupLevel = {
    BEGINNER: 'beginner',
    INTERMEDIATE: 'intermediate',
    ADVANCED: 'advanced'
};

export const GroupService = {
    // Create a new group
    createGroup: async (name, level, adminId) => {
        try {
            const groupData = {
                name,
                level,
                adminId,
                memberIds: [],
                createdAt: serverTimestamp()
            };
            const docRef = await addDoc(collection(db, "groups"), groupData);
            return docRef.id;
        } catch (error) {
            console.error("Error creating group:", error);
            throw error;
        }
    },

    // Get all groups as a stream (callback)
    getGroupsStream: (callback) => {
        const q = query(collection(db, "groups"));
        return onSnapshot(q, (snapshot) => {
            const groups = snapshot.docs.map(doc => ({
                id: doc.id,
                ...doc.data()
            }));
            callback(groups);
        });
    },

    // Get groups managed by a specific admin
    getGroupsByAdminStream: (adminId, callback) => {
        const q = query(collection(db, "groups"), where("adminId", "==", adminId));
        return onSnapshot(q, (snapshot) => {
            const groups = snapshot.docs.map(doc => ({
                id: doc.id,
                ...doc.data()
            }));
            callback(groups);
        });
    },

    // Delete a group and remove references from members
    deleteGroup: async (groupId, memberIds = []) => {
        try {
            await deleteDoc(doc(db, "groups", groupId));

            // Remove assignedGroupId from all members
            // Note: Batching would be better for large groups, but doing individually for now
            // mirroring existing Flutter logic which implies some cleanup
            if (memberIds && memberIds.length > 0) {
                const batchPromises = memberIds.map(memberId =>
                    updateDoc(doc(db, "users", memberId), {
                        assignedGroupId: null
                    })
                );
                await Promise.all(batchPromises);
            }
        } catch (error) {
            console.error("Error deleting group:", error);
            throw error;
        }
    },

    // Add a member to a group
    addMemberToGroup: async (groupId, userId) => {
        try {
            // Add user ID to group's memberIds array
            await updateDoc(doc(db, "groups", groupId), {
                memberIds: arrayUnion(userId)
            });

            // Update user's assignedGroupId
            await updateDoc(doc(db, "users", userId), {
                assignedGroupId: groupId
            });
        } catch (error) {
            console.error("Error adding member to group:", error);
            throw error;
        }
    },

    // Remove a member from a group
    removeMemberFromGroup: async (groupId, userId) => {
        try {
            // Remove user ID from group's memberIds array
            await updateDoc(doc(db, "groups", groupId), {
                memberIds: arrayRemove(userId)
            });

            // Update user's assignedGroupId to null
            await updateDoc(doc(db, "users", userId), {
                assignedGroupId: null
            });
        } catch (error) {
            console.error("Error removing member from group:", error);
            throw error;
        }
    },

    // Get members of a specific group
    getGroupMembersStream: (groupId, callback) => {
        // Query users collection checking all possible group ID fields
        const q = query(
            collection(db, "users"),
            or(
                where("assignedGroupId", "==", groupId),
                where("groupId", "==", groupId),
                where("group", "==", groupId)
            )
        );
        return onSnapshot(q, (snapshot) => {
            const users = snapshot.docs.map(doc => ({
                id: doc.id,
                ...doc.data()
            }));
            callback(users);
        });
    }
};
